pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+	var auths dockerConfigFile
+
+	raw, err := ioutil.ReadFile(path)
+	if os.IsNotExist(err) {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+		return auths, nil
+	}
+
+	if err = json.Unmarshal(raw, &auths); err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+	}
+
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func modifyJSON(registry, username, password, path, action string) error {
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	switch action {
+	case addAuth:
+		if len(auths.CredHelpers) > 0 {
+			if err = setAuthToCredHelper(registry, username, password); err != nil {
+				return err
+			}
+			break
+		}
+
+		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		newCreds := dockerAuthConfig{Auth: creds}
+		auths.AuthConfigs[registry] = newCreds
+	case deleteAuth:
+		// First try cred helpers.
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			if err = deleteAuthFromCredHelper(ch, registry); err != nil {
+				return err
+			}
+			break
+		}
+
+		if _, ok := auths.AuthConfigs[registry]; ok {
+			delete(auths.AuthConfigs, registry)
+		} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+			delete(auths.AuthConfigs, normalizeRegistry(registry))
+		} else {
+			return ErrNotLoggedIn
+		}
+	}
+
+	newData, err := json.MarshalIndent(auths, "", "\t")
+	if err != nil {
+		return errors.Wrapf(err, "error marshaling JSON %q", path)
+	}
+
+	if err = ioutil.WriteFile(path, newData, 0755); err != nil {
+		return errors.Wrapf(err, "error writing to file %q", path)
+	}
+	return nil
+}
+
+func getAuthFromCredHelper(credHelper, registry string) (string, string, error) {
+	registry = convertToHostname(registry)
+	helperName := fmt.Sprintf("docker-credential-%s", credHelper)
+	p := helperclient.NewShellProgramFunc(helperName)
+	creds, err := helperclient.Get(p, registry)
+	if err != nil {
+		return "", "", err
+	}
+	return creds.Username, creds.Secret, nil
+}
+
+func setAuthToCredHelper(registry, username, password string) error {
+	registry = convertToHostname(registry)
+	helperName := fmt.Sprintf("docker-credential-%s", registry)
+	p := helperclient.NewShellProgramFunc(helperName)
+	creds := &credentials.Credentials{
+		ServerURL: registry,
+		Username:  username,
+		Secret:    password,
+	}
+	return helperclient.Store(p, creds)
+}
+
+func deleteAuthFromCredHelper(credHelper, registry string) error {
+	registry = convertToHostname(registry)
+	helperName := fmt.Sprintf("docker-credential-%s", registry)
+	p := helperclient.NewShellProgramFunc(helperName)
+	return helperclient.Erase(p, registry)
+}
+
+// findAuthentication looks for auth of registry in path
+func findAuthentication(registry, path string) (string, string, error) {
+	registry = DockerHostNameToRegistry(registry)
```

Comment:
The caller has already done this mapping.

Reply:
fixed.